firebase-firestore/src/main/java/com/google/firebase/firestore/model/protovalue/ObjectValue.java

```diff
@@ -0,0 +1,237 @@
+        if (!isType(value, TYPE_ORDER_OBJECT)) {
+          return null;
+        }
+      }
+      value = value.getMapValue().getFieldsOrDefault(fieldPath.getLastSegment(), null);
+      return value != null ? FieldValue.of(value) : null;
+    }
+  }
+
+  /** Recursively extracts the FieldPaths that are set in this ObjectValue. */
+  public FieldMask getFieldMask() {
+    return extractFieldMask(internalValue.getMapValue());
+  }
+
+  private FieldMask extractFieldMask(MapValue value) {
+    Set<FieldPath> fields = new HashSet<>();
+    for (Map.Entry<String, Value> entry : value.getFieldsMap().entrySet()) {
+      FieldPath currentPath = FieldPath.fromSingleSegment(entry.getKey());
+      if (isType(entry.getValue(), TYPE_ORDER_OBJECT)) {
+        FieldMask nestedMask = extractFieldMask(entry.getValue().getMapValue());
+        Set<FieldPath> nestedFields = nestedMask.getMask();
+        if (nestedFields.isEmpty()) {
+          // Preserve the empty map by adding it to the FieldMask.
+          fields.add(currentPath);
+        } else {
+          // For nested and non-empty ObjectValues, add the FieldPath of the leaf nodes.
+          for (FieldPath nestedPath : nestedFields) {
+            fields.add(currentPath.append(nestedPath));
+          }
+        }
+      } else {
+        fields.add(currentPath);
+      }
+    }
+    return FieldMask.fromSet(fields);
+  }
+
+  /** Creates a ObjectValue.Builder instance that is based on the current value. */
+  public ObjectValue.Builder toBuilder() {
+    return new Builder(this);
+  }
+
+  /** An ObjectValue.Builder provides APIs to set and delete fields from an ObjectValue. */
+  public static class Builder {
+
+    /** The existing data to mutate. */
+    private ObjectValue baseObject;
+
+    /**
+     * A nested map that contains the accumulated changes in this builder. Values can either be
+     * `Value` protos, `Map<String, Object>` values (to represent additional nesting) or `null` (to
+     * represent field deletes).
+     */
+    private Map<String, Object> overlayMap;
+
+    Builder(ObjectValue baseObject) {
+      this.baseObject = baseObject;
+      this.overlayMap = new HashMap<>();
+    }
+
+    /**
+     * Sets the field to the provided value.
+     *
+     * @param path The field path to set.
+     * @param value The value to set.
+     * @return The current Builder instance.
+     */
+    public Builder set(FieldPath path, Value value) {
+      hardAssert(!path.isEmpty(), "Cannot set field for empty path on ObjectValue");
+      setOverlay(path, value);
+      return this;
+    }
+
+    /**
+     * Removes the field at the specified path. If there is no field at the specified path nothing
+     * is changed.
+     *
+     * @param path The field path to remove
+     * @return The current Builder instance.
+     */
+    public Builder delete(FieldPath path) {
+      hardAssert(!path.isEmpty(), "Cannot delete field for empty path on ObjectValue");
+      setOverlay(path, null);
```

Comment:
`null` is pretty overloaded. I wonder if it would be helpful to store a sentinel in here instead.

Something like this up top:

```java
private static final Object DELETED = new Object();
```

And then this would read as
```java
setOverlay(path, DELETED);
```

I don't think this changes any of the handling below, so probably not worth it, but it has some visual appeal.

Feel free to ignore.

Reply:
I am leaving this as is for now. The reason is the DELETE sentinel would need to of type `Value` so I can pass it to `setOverlay`. That makes my `instanceof` checks more complicated, as I now have to differentiate between user-provided values and the sentinel.